control-plane/api/v1alpha1/terminatinggatewayservice_types.go

```diff
@@ -0,0 +1,147 @@
+	// TaggedAddresses specifies the tagged addresses.
+	TaggedAddresses map[string]string `json:"taggedAddresses,omitempty"`
+	// NodeMeta specifies arbitrary KV metadata pairs for filtering purposes.
+	NodeMeta map[string]string `json:"nodeMeta,omitempty"`
+	// ServiceID specifies to register a service. If ID is not provided, it will be defaulted to the value of the ServiceName property.
+	ServiceID string `json:"serviceID,omitempty"`
+	// ServiceName specifies the logical name of the service.
+	ServiceName string `json:"serviceName,omitempty"`
+	// ServiceAddress specifies the address of the service. If not provided, the agent's address is used as the address for the service during DNS queries.
+	ServiceAddress string `json:"serviceAddress,omitempty"`
+	// ServiceTags specifies a list of tags to assign to the service. These tags can be used for later filtering and are exposed via the APIs.
+	ServiceTags []ServiceTag `json:"serviceTags,omitempty"`
+	// SericeMeta specifies arbitrary KV metadata linked to the service instance.
+	ServiceMeta map[string]string `json:"serviceMeta,omitempty"`
+	// ServicePort specifies the port of the service.
+	ServicePort int `json:"servicePort,omitempty"`
+	// ServiceEnableTagOverride specifies to disable the anti-entropy feature for this service's tags.
+	// If set to true then external agents can update this service in the catalog and modify the tags.
+	ServiceEnableTagOverride bool `json:"serviceEnableTagOverride,omitempty"`
+}
+
+type ServiceTag string
+
+// TerminatingGatewayServiceStatus defines the observed state of TerminatingGatewayService.
+type TerminatingGatewayServiceStatus struct {
+	// Important: Run "make" to regenerate code after modifying this file
+	// ServiceInfoRefStatus shows information about the service.
+	ServiceInfoRef *ServiceInfoRefStatus `json:"serviceInfoRef,omitempty"`
+	// Conditions indicate the latest available observations of a resource's current state.
+	// +optional
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	Conditions Conditions `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type"`
+	// LastSyncedTime is the last time the resource successfully synced with Consul.
+	// +optional
+	LastSyncedTime *metav1.Time `json:"lastSyncedTime,omitempty" description:"last time the condition transitioned from one status to another"`
+}
+
+type ServiceInfoRefStatus struct {
+	ServiceName string `json:"serviceInfo,omitempty"`
+	PolicyName  string `json:"service,omitempty"`
+}
+
+func (tas *TerminatingGatewayService) ServiceInfo() *CatalogService {
+	return tas.Spec.Service
+}
+func (tas *TerminatingGatewayService) ServiceInfoRef() *ServiceInfoRefStatus {
+	return tas.Status.ServiceInfoRef
+}
+func (tas *TerminatingGatewayService) KubeKind() string {
+	return TerminatingGatewayServiceKubeKind
+}
+func (tas *TerminatingGatewayService) KubernetesName() string {
+	return tas.ObjectMeta.Name
+}
+func (tas *TerminatingGatewayService) Validate() error {
```

Comment:
Can we add a test for this function? It should have a few invalid cases and a valid case.

Reply:
Yes. Done